superset/dao/base.py

```diff
@@ -0,0 +1,108 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from typing import Dict, List, Optional
+
+from flask import current_app
+from flask_appbuilder.models.sqla import Model
+from flask_appbuilder.security.sqla.models import User
+from sqlalchemy.exc import SQLAlchemyError
+
+from superset.commands.exceptions import (
+    CreateFailedError,
+    DeleteFailedError,
+    OwnersNotFoundValidationError,
+    UpdateFailedError,
+)
+from superset.extensions import db
+
+
+def generic_create(model_cls: Model, properties: Dict, commit=True) -> Optional[Model]:
```

Comment:
Don't pass in the class here. Instead, the caller will have created whatever instance he wants and then should pass it in. You can then combine your create / update into a single `upsert` method. Also, you can call this something like `def upsert(model, commit=True)`

Reply:
We can do that and end up with less code. Yet I tend to think it's more readable this way (and easier to reason), DAO create methods call `generic_create`, update methods call `generic_update`, also Exceptions follow the same pattern.